---
title: "Expression profiling by high throughput sequencing: RNA-seq of longissimus
  dorsi muscles in Japanese black (Wagyu) and Chinese Red Steppes cattle"
author: "Linda Lillian"
date: "15 jan 2023"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
**METHODS**

**Linux tools used**
fasqc v0.11.9, multiqc v1.14, hisat2 v2.1.0, Samtools V1.10, Stringtie V2.2.1, gffcompare V0.12.6.

**R packages used**
GenomeInfoDbData v1.2.7, ggplot2 v3.4.0 , ballgown v2.26.0, genefilter v1.76.0, RSkittleBrewer v1.1, devtools v2.4.5, dplyr v1,0.10, ggrepel v0.9.2, pheatmap v1.0.12, gplots v3.1.3, GenomicRanges v1.46.1 and viridis v0.6.2

**Procedure**
Fastqc tool was used for quality control of the raw reads.
Hisat2 tool was used to index build the reference and mapping the paired reads to the indexed reference to obtain SAM files
Samtools were used for sorting the SAM files and converted to BAM files.
The BAM files were used for transcript assembly on the reference annotated gtf file using stringtie tool and merged together.
Gffcompare was used to check the precision and sensitivity statistics of the assembled transcripts.
Stringtie tool was used for abundance estimation and the files produced in this step then used as an input together with the phenotypic data for differential expression in R using the ballgown package.

**RESULTS**

**Loading Packages in R**
```{r message=FALSE, warning=FALSE}
#load the libraries
library(GenomeInfoDbData)
library(ggplot2)
library(ballgown)
library(genefilter)
library(RSkittleBrewer)
library(devtools)
library(dplyr)
library(ggrepel)
library(pheatmap)
library(gplots)
library(GenomicRanges)
library(viridis)
```
**Creating phenotype data**
```{r}
pheno_data<-data.frame(
  Sample= c("SRR13107018", "SRR13107019", "SRR13107020", "SRR13107021", "SRR13107022", "SRR13107023"),
  Breed = c("Japanese black cattle", "Japanese black cattle", "Japanese black cattle", "Chinese Red Steppes cattle", "Chinese Red Steppes cattle", "Chinese Red Steppes cattle"))
pheno_data
```

**Load the expression data using ballgown**
```{r message=FALSE}
bg_chrX <- ballgown(dataDir="data/ballgown",samplePattern="SRR",pData=pheno_data)

```
Number of transcripts
```{r}
bg_chrX

```

**filtering out transcripts with low variance in order done to remove some genes that have few counts. Filtering improves the statistical power of differential expression analysis.**

```{r}
bg_chrX_filt<- subset(bg_chrX,"rowVars(texpr(bg_chrX))>1",genomesubset=TRUE)

```
Number of trancsripts remaining after filtering
```{r}
bg_chrX_filt
```
**Differentially expressed transcripts in the differnt cattle breeds**
```{r}
de_transcripts <- stattest(bg_chrX_filt,feature="transcript",covariate="Breed",getFC=TRUE,meas="FPKM")
head(de_transcripts)
```

**Differentially expressed genes in the differnt cattle breeds**
```{r}
de_genes <- stattest(bg_chrX_filt,feature="gene",covariate="Breed", getFC=TRUE, meas="FPKM")
head(de_genes)
```
#adding identifiers
```{r}
#add identifiers
de_transcripts = data.frame(geneNames=ballgown::geneNames(bg_chrX_filt), geneIDs=ballgown::geneIDs(bg_chrX_filt), de_transcripts)
```

Get gene names and check for isoforms
```{r}
# Let's test on genes
de_genes <- stattest(bg_chrX_filt,feature="gene",covariate="Breed", getFC=TRUE, meas="FPKM")
bg_filt_table=texpr(bg_chrX_filt,'all')
gene_names=unique(bg_filt_table[,9:10])
features=de_genes$id
mapped_gene_names=vector()
for (i in features) 
{  query=gene_names%>%filter(gene_id==i & gene_name != '.') ; n_hit=dim(query)[1]; if (n_hit==1) {mapped_gene_names=append(mapped_gene_names,query$gene_name[[1]]) } else
{mapped_gene_names=append(mapped_gene_names,'.') }    
}
```

#Add the mapped gene names to the de genes table
```{r}
#add the mapped gene names to the de genes table
de_genes$gene_name <- mapped_gene_names
de_genes <- de_genes[, c('feature','gene_name','id','fc','pval','qval')]
head(de_genes)
```

Adding a column for log2 for both trancsripts and genes
```{r}
de_genes[,"log2fc"] <- log2(de_genes[,"fc"])
de_transcripts[,"log2fc"] <- log2(de_transcripts[,"fc"])
```

Arrange the results from the smallest P value to the largest
```{r}
de_transcripts = arrange(de_transcripts,pval)
de_genes = arrange(de_genes,pval)
```

**Transcripts and that are not significantly expressed**

```{r}
table(de_transcripts$qval <0.05)
de_transcripts %>%filter(qval <0.05)
```
**Transcripts and that are significantly expressed**
```{r}
table(de_genes$qval <0.05)
de_genes %>%filter(qval <0.05)
```

**Sub-setting transcripts and genes that are significantly at qval <0.05**
```{r}
subset_transcripts <- subset(de_transcripts,de_transcripts$qval<0.05)
subset_genes <- subset(de_genes,de_genes$qval<0.05)
```


#gene expression for a isoforms of gene PARP11
```{r warning=F}
myplot=plotTranscripts(ballgown::geneIDs(bg_chrX)[ballgown::geneNames(bg_chrX) == "PARP11"], bg_chrX, main=c('Gene PARP11 in sample SRR13107019'), sample=c('SRR13107018'))
print(myplot)

```
No isoforms

**Differentially expressed genes**
```{r warning=FALSE}
#de_genes$diffexpressed[de_genes$log2fc < -0.6 & de_genes$qval < 0.05] <- "DOWN"
de_genes$diffexpressed <- "NO"
de_genes$diffexpressed[de_genes$log2fc > 1 & de_genes$pval < 0.05] <- "UP"
de_genes$diffexpressed[de_genes$log2fc < -1 & de_genes$pval < 0.05] <- "DOWN"
de_genes$delabel <- NA
de_genes$delabel[de_genes$diffexpressed != "NO"] <- de_genes$id[de_genes$diffexpressed != "NO"]

options(ggrepel.max.overlaps = Inf)

volcano=ggplot(data=de_genes, aes(x=log2fc, y=-log10(pval), col=diffexpressed, label=delabel)) +
  geom_point() + 
  theme_minimal() +
  geom_text_repel() +
  scale_color_manual(values=c("blue", "black", "red")) +
  geom_vline(xintercept=c(-0.8, 0.8), col="red") +
  theme(text=element_text(size=20))
print(volcano)

```







```{r}
de_transcripts$mean <- rowMeans(texpr(bg_chrX_filt))
maplot=ggplot(de_transcripts, aes(log2(mean), log2(fc), colour = qval<0.05)) +
  scale_color_manual(values=c("#999999", "#FF0000")) +
  geom_point() +
  theme(legend.text=element_text(size=20),legend.title=element_text(size=20)) +
  theme(axis.text=element_text(size=20),axis.title=element_text(size=20)) +
  geom_hline(yintercept=0)
print(maplot)
```

#heatmap_clustered fpkm values 
```{r}
fpkm = gexpr(bg_chrX_filt)
fpkm = log2(fpkm+1)
g_ids=subset_genes$id
hits= which (de_genes$id %in% g_ids)
hit_frame=fpkm[hits,]
row.names(hit_frame) <- g_ids
heatmap_image=pheatmap(hit_frame)
print(heatmap_image)
```

**heatmap_unclustered fpkm values**
```{r}
heatmap_image=pheatmap(hit_frame,cluster_rows = F,cluster_cols=F)
print(heatmap_image)
```





